src/components/lib/crt/crt.c

```diff
@@ -105,6 +105,319 @@ crt_chkpt_restore(struct crt_chkpt *chkpt, struct crt_comp *c)
+	}
+
+	/* initialize an MPK NS for new */
+	for(i = 0 ; i < CRT_VAS_NUM_NAMES ; i++) {
+		new->mpk_names[i].reserved = 1;
+		new->mpk_names[i].allocated = 0;
+	}
+
+	return 0;
+
+}
+
+/*
+ * Create a new vas namespace from the names "left over" in
+ * `existing`, i.e. those that have not been `crt_ns_vas_alloc_in`ed
+ * 
+ * Automatically alias all names from existing into new
+ * 
+ * Return values:
+ *   0: success
+ *  -1: new is null/not allocated correctly, or initialization fails
+ *  -2: new already has allocations
+ */
+int crt_ns_vas_split(struct crt_ns_vas *new, struct crt_ns_vas *existing, struct crt_ns_asid *asids)
+{
+	int i;
+	int cons_ret;
+
+	/* verify that `new` has no existing allocations */
+	if(new == NULL || new->names == NULL) {
+		return -1;
+	}
+	for(i = 0 ; i < CRT_VAS_NUM_NAMES ; i++) {
+		if(new->names[i].allocated == 1) {
+			return -2;
+		} 
+	}
+
+	if(crt_ns_vas_init(new, asids)) {
+		return -1;
+	}
+
+	for(i = 0 ; i < CRT_VAS_NUM_NAMES ; i++) {
+		/* if a name is allocated or aliased in existing, the component there should automatically be aliased into new */
+		/* by default via init everything else will go to:
+		 * 		reserved  = 1
+		 *      allocated = 0
+		 *      aliased   = 0
+		 */
+		if(existing->names[i].allocated || existing->names[i].aliased) {
+			new->names[i].reserved = 0;
+			new->names[i].aliased = 1;
+			new->names[i].comp = existing->names[i].comp;
+
+			cons_ret = cos_cons_into_shared_pgtbl(cos_compinfo_get(new->names[i].comp->comp_res), new->top_lvl_pgtbl);
+			if(cons_ret != 0) {
+				printc("cons failed: %d\n", cons_ret);
+				assert(0);
+			}
+
+		}
+		/* if a name is reserved (but not allocated) in existing, it should no longer be reserved in existing 
+		 * NOTE: this means no further allocations can be made in existing
+		 */
+		if(existing->names[i].reserved == 1 && existing->names[i].allocated == 0) {
+			existing->names[i].reserved = 0;
+		}
+	}
+
+	/* initialize the mpk namespace within new */
+	for(i = 0 ; i < CRT_MPK_NUM_NAMES ; i++) {
+		if(existing->mpk_names[i].allocated == 1) {
+			new->mpk_names[i].reserved = 0;
+		}
+		if(existing->mpk_names[i].reserved == 1 && existing->mpk_names[i].allocated == 0) {
+			existing->mpk_names[i].reserved = 0;
+		}
+	}
+	new->parent = existing;
+
+	return 0;
+}
+
+
+/* 
+ * helper function:
+ * returns the first available MPK name within vas, or -1 if none available
+ */
+int
+crt_mpk_available_name(struct crt_ns_vas *vas)
+{
+	/* we're looking for an MPK name that's not allocated */
+	int i = 0;
+
+	while(i < CRT_MPK_NUM_NAMES) {
+		if(vas->mpk_names[i].reserved == 1 && vas->mpk_names[i].allocated == 0) {
+			return i;
+		}
+		i++;
+	}
+
+	return -1;
+}
+
+/*
+ * VAS name mapping/allocation. This function allocates `c` into `vas`
+ * by tracking which names (MPK & VAS) are dedicated to `c`. A
+ * component can only be allocated into a *single* vas.
+ */
+int crt_ns_vas_alloc_in(struct crt_ns_vas *vas, struct crt_comp *c)
```

Comment:
Do we need this function? Is the next sufficient where we allocate the component into the vas?